server/sonar-server/src/main/java/org/sonar/server/computation/step/PersistDuplicationMeasuresStep.java

```diff
@@ -0,0 +1,142 @@
+import org.sonar.core.component.ComponentKeys;
+import org.sonar.core.measure.db.MeasureDto;
+import org.sonar.core.measure.db.MetricDto;
+import org.sonar.core.persistence.DbSession;
+import org.sonar.core.persistence.MyBatis;
+import org.sonar.server.computation.ComputationContext;
+import org.sonar.server.db.DbClient;
+
+import javax.annotation.Nullable;
+
+import java.util.List;
+
+public class PersistDuplicationMeasuresStep implements ComputationStep {
+
+  private final DbClient dbClient;
+
+  public PersistDuplicationMeasuresStep(DbClient dbClient) {
+    this.dbClient = dbClient;
+  }
+
+  @Override
+  public String[] supportedProjectQualifiers() {
+    return new String[] {Qualifiers.PROJECT};
+  }
+
+  @Override
+  public void execute(ComputationContext context) {
+    DbSession session = dbClient.openSession(true);
+    try {
+      MetricDto duplicationMetric = dbClient.metricDao().selectByKey(session, CoreMetrics.DUPLICATIONS_DATA_KEY);
+      int rootComponentRef = context.getReportMetadata().getRootComponentRef();
+      recursivelyProcessComponent(session, context, null, rootComponentRef, duplicationMetric);
+      session.commit();
+    } finally {
+      MyBatis.closeQuietly(session);
+    }
+  }
+
+  private void recursivelyProcessComponent(DbSession session, ComputationContext context, @Nullable Integer parentComponentRef, int componentRef, MetricDto duplicationMetric) {
+    BatchReportReader reportReader = context.getReportReader();
+    String branch = context.getReportMetadata().getBranch();
+    BatchReport.Component component = reportReader.readComponent(componentRef);
+    List<BatchReport.Duplication> duplications = reportReader.readComponentDuplications(componentRef);
+    if (!duplications.isEmpty() && parentComponentRef != null) {
+      saveDuplications(session, reportReader, reportReader.readComponent(parentComponentRef), component, duplications, duplicationMetric, branch);
+    }
+
+    for (Integer childRef : component.getChildRefList()) {
+      recursivelyProcessComponent(session, context, componentRef, childRef, duplicationMetric);
+    }
+  }
+
+  private void saveDuplications(DbSession session, BatchReportReader reportReader, BatchReport.Component parentComponent, BatchReport.Component component,
+                                List<BatchReport.Duplication> duplications, MetricDto duplicationMetric, @Nullable String branch) {
+    String duplicationXml = createXmlDuplications(reportReader, parentComponent, component.getPath(), duplications, branch);
+    MeasureDto measureDto = new MeasureDto()
+      .setMetricId(duplicationMetric.getId())
+      .setData(duplicationXml)
+      .setComponentId(component.getId())
+      .setSnapshotId(component.getSnapshotId());
+    dbClient.measureDao().insert(session, measureDto);
+  }
+
+  private String createXmlDuplications(BatchReportReader reportReader, BatchReport.Component parentComponent, String componentPath,
```

Comment:
I understand, but I think it's the responsability of this class to persist measures of duplications, and these measures are stored as XML.